service/history/timer_queue_standby_task_executor.go

```diff
@@ -472,6 +472,71 @@ func (t *timerQueueStandbyTaskExecutor) executeWorkflowExecutionTimeoutTask(
 		),
 	)
 }
+func (t *timerQueueStandbyTaskExecutor) executeStateMachineTimerTask(
+	ctx context.Context,
+	timerTask *tasks.StateMachineTimerTask,
+) error {
+	actionFn := func(
+		ctx context.Context,
+		wfContext workflow.Context,
+		mutableState workflow.MutableState,
+	) (any, error) {
+		timers, err := t.executeStateMachineTimers(
+			ctx,
+			mutableState,
+			func(node *hsm.Node, task hsm.Task) error {
+				if task.Concurrent() {
+					//nolint:revive // concurrent tasks implements hsm.ConcurrentTask interface
+					concurrentTask := task.(hsm.ConcurrentTask)
+					return concurrentTask.Validate(node)
+				}
+				// If the task is expired and still valid in the standby queue,
+				// then the state machine is stale.
+				return consts.ErrTaskRetry
+			},
+		)
+		if err != nil {
+			if errors.Is(err, consts.ErrTaskRetry) {
+				// This handles the ErrTaskRetry error returned by executeStateMachineTimers.
+				return getHistoryResendInfo(mutableState)
+			}
+			return nil, err
+		}
+
+		// We haven't done any work, return without committing.
```

Comment:
I realized this condition is wrong. There's a slice of tasks in each of these `StateMachineTimers`.

Reply:
I can fix in a followup PR.

Reply:
I looked at this again and it's fine because the only reason we won't process all timers in the slice for a single duration is when one of them fails and in that case we don't need to commit.